### PR TITLE
set correct remaining available discount amount when applying fixed p…

### LINKED ISSUE #30952: [Cart Rule] Incorrect discount amount applied for bundle products containing more than one option
### Preconditions (*)

1. Magento 2.4 (probably older versions too)
2. Affects bundle products fixed cart discounts

Note: Because the `$availableDiscountAmount` is set to zero after processing the first item the other items will not have any discount applied.

### Steps to reproduce (*)

1. Install Magento 2.4 with sample data
2. Create cart rule with fixed discount amount for whole cart (Ex: 50$)
3. Add a bundle product to cart (Example: http://local.magento2.com/sprite-yoga-companion-kit.html)
4. Navigate to checkout/cart page and check discount. 

### Expected result (*)

1. Cart discount should be 50$
image

### Actual result (*)

1. Cart discount is 18.85$
image

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [X] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

Fixed in: #30953

### PR DESCRIPTION
**Pull request by sky-hub**
…rice cart rules to bundle products

### Fixed Issues (if relevant)

1.  fixes https://github.com/magento/magento2/issues/30952

### Manual testing scenarios (*)

1. Install Magento 2.4 with sample data
2. Create cart rule with fixed discount amount for whole cart (Ex: 50$)
3. Add a bundle product to cart (Example: http://local.magento2.com/sprite-yoga-companion-kit.html)
4. Navigate to checkout/cart page and check discount.
5. Expected result: 50$, actual: 18.85$

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by eduard13** [comment]
@magento run all tests

**Review by eduard13 — CHANGES_REQUESTED** [review]
Hi @sky-hub, thank you for the following fix. According to the [Definition of Done](https://devdocs.magento.com/contributor-guide/contributing_dod.html#automated-tests), all changes have to be covered by some automated tests. Would you be kind covering the fixed case by a test? 
You may write a simple [MFTF test](https://devdocs.magento.com/mftf/docs/introduction.html) that will pass through the Precondition steps and will check the cart discount.

Thank you.

**Comment by sky-hub** [comment]
@eduard13, sure. I will get back with a comment when tests are implemented.

Cheers

**Comment by sky-hub** [comment]
@eduard13 I had a look in the MTFT testing and I ran into a dead end.

It seems there is already a MTFT Test for bundle products in `app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartPriceRuleForBundleProductTest.xml` but apparently even this test is skipped due to some issue: 
`<skip><issueId value="MQE-2288" /></skip>`.

At this point I do not know if I should update this test, and if I do update it what is the reason behind that issue that marks the test as skipped. Do you think you can help here?

PS: Please note that my experience with MTFT is very limited.

**Comment by eduard13** [comment]
@sky-hub ahh, got your point. Based on the commit's message, it looks like there were some issues running the test on Chrome 85 version. As far as I can see, the existing test applies on `Percent of product price discount`, which in my opinion should also be the case for this PR (I haven't checked this case), however you may create a new one that checks the `Fixed amount discount` test case (having the same test case as the issue states). 

What do you think?
Thank you.

**Comment by eduard13** [comment]
@magento run Integration Tests, Functional Tests B2B

**Comment by sky-hub** [comment]
@eduard13 , sounds good to me.
I'll implement a test as you suggested and get back to you when it's done

**Comment by sky-hub** [comment]
@eduard13 , test implemented.
Please let me know if there is anything else I can help with.

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        <fillField selector="{{AdminCartPriceRulesFormSection.fromDate}}" userInput="{$yesterdayDate}" stepKey="fillFromDate"/>
+        <click selector="{{AdminCartPriceRulesFormSection.actionsHeader}}" stepKey="clickToExpandActions"/>
+        <selectOption selector="{{AdminCartPriceRulesFormSection.apply}}" userInput="Fixed amount discount for whole cart" stepKey="selectActionType"/>
+        <fillField selector="{{AdminCartPriceRulesFormSection.discountAmount}}" userInput="4.99" stepKey="fillDiscountAmount"/>
+        <click selector="{{AdminCartPriceRulesFormSection.save}}" stepKey="clickSaveButton"/>
+        <see selector="{{AdminCartPriceRulesSection.messages}}" userInput="You saved the rule." stepKey="seeSuccessMessage"/>
+
+        <!-- Add the first product to the cart -->
+        <amOnPage url="$$createBundleProduct.sku$$.html" stepKey="goToProductPage1"/>
+        <waitForPageLoad stepKey="waitForProductPageLoad1"/>
```
Comment (eduard13) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml]: Please try to re-use the existing ActionGroups as much as possible. In this case, we have the `StorefrontOpenProductPageActionGroup` that receives the product url as a parameter.

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        <selectOption selector="{{AdminCartPriceRulesFormSection.websites}}" userInput="Main Website" stepKey="selectWebsites"/>
+        <selectOption selector="{{AdminCartPriceRulesFormSection.customerGroups}}" userInput="NOT LOGGED IN" stepKey="selectCustomerGroup"/>
+        <generateDate date="-1 day" format="m/d/Y" stepKey="yesterdayDate"/>
+        <fillField selector="{{AdminCartPriceRulesFormSection.fromDate}}" userInput="{$yesterdayDate}" stepKey="fillFromDate"/>
+        <click selector="{{AdminCartPriceRulesFormSection.actionsHeader}}" stepKey="clickToExpandActions"/>
+        <selectOption selector="{{AdminCartPriceRulesFormSection.apply}}" userInput="Fixed amount discount for whole cart" stepKey="selectActionType"/>
+        <fillField selector="{{AdminCartPriceRulesFormSection.discountAmount}}" userInput="4.99" stepKey="fillDiscountAmount"/>
+        <click selector="{{AdminCartPriceRulesFormSection.save}}" stepKey="clickSaveButton"/>
+        <see selector="{{AdminCartPriceRulesSection.messages}}" userInput="You saved the rule." stepKey="seeSuccessMessage"/>
```
Comment (eduard13) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml]: I'd create a new ActionGroup that is going to create a new Cart Rule, based on our needs. As arguments, you'll have all the hardcoded inputs from this test:
- ruleName
- websites
- customerGroups
- apply
- discountAmount

This way, we'll be able to re-use this action group in some other tests.
Thank you.

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        <!--Click "Customize and Add to Cart" button-->
+        <click selector="{{StorefrontBundledSection.addToCart}}" stepKey="clickCustomize"/>
+
+        <!-- Select two products -->
+        <click stepKey="selectProduct1" selector="{{StorefrontBundledSection.productCheckbox('1','1')}}"/>
+        <click stepKey="selectProduct2" selector="{{StorefrontBundledSection.productCheckbox('2','1')}}"/>
+
+        <!--Click "Add to Cart" button-->
+        <click selector="{{StorefrontBundleProductActionSection.addToCartButton}}" stepKey="clickAddBundleProductToCart"/>
+        <waitForPageLoad time="30" stepKey="waitForAddBundleProductPageLoad"/>
```
Comment (eduard13) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml]: Could you please double check if you're able to use any of existing ActionGroups that is supposed to add a Bundle product to cart?
Thank you.

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+        <!--Click "mini cart" icon-->
+        <actionGroup ref="StorefrontOpenCartFromMinicartActionGroup" stepKey="openCart"/>
+        <waitForPageLoad stepKey="waitForDetailsOpen"/>
```
Comment (eduard13) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml]: Same here: `StorefrontCartPageOpenActionGroup`

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    <test name="StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest">
+        <annotations>
+            <features value="SalesRule"/>
+            <stories value="Fixed Amount Cart Price Rule"/>
+            <title value="Checking Fixed Amount Cart Price Rule"/>
+            <description value="Checking Fixed Amount Cart Price Rule"/>
+            <severity value="AVERAGE"/>
+            <group value="SalesRule"/>
+        </annotations>
+
```
Comment (eduard13) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml]: Please remove all the empty rows.
```suggestion
```

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            </createData>
+            <createData entity="ApiBundleLink" stepKey="linkOptionToProduct4">
+                <requiredEntity createDataKey="createBundleProduct"/>
+                <requiredEntity createDataKey="createBundleOption1_2"/>
+                <requiredEntity createDataKey="simpleProduct4"/>
+            </createData>
+
+            <actionGroup ref="AdminLoginActionGroup" stepKey="loginAsAdmin"/>
+
+            <!-- Make Attribute 'sku' accessible for Promo Rule Conditions -->
```
Comment (eduard13) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml]: Also, remove all the comments please, they are unnecessary.

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  */
+-->
+<tests xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/testSchema.xsd">
+    <test name="StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest">
+        <annotations>
+            <features value="SalesRule"/>
+            <stories value="Fixed Amount Cart Price Rule"/>
+            <title value="Checking Fixed Amount Cart Price Rule"/>
+            <description value="Checking Fixed Amount Cart Price Rule"/>
```
Comment (eduard13) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml]: Can we extend the title and description here by adding a more meaningful information on what does this test do?

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml [hunk 8]
```diff
@@ -0,0 +1,146 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+ /**
+  * Copyright © Magento, Inc. All rights reserved.
+  * See COPYING.txt for license details.
+  */
+-->
+<tests xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/testSchema.xsd">
+    <test name="StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest">
```
Comment (eduard13) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartFixedAmountWholeCartDiscountForBundleProductTest.xml]: What do you think if we rename the test as following?
```suggestion
    <test name="StorefrontAssertFixedCartDiscountAmountForBundleProductTest">
```

**Review by eduard13 — CHANGES_REQUESTED** [review]
Hi @sky-hub, thank you for the test coverage. I've left couple of comments, could you please review them?
Feel free to reach me if additional questions arise.

Thanks.

**Comment by eduard13** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by eduard13** [comment]
Additionally, could you please review the failed Functional Tests, as they might be related to your changes?
Thank you.

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
Functional Tests CE

**Comment by sky-hub** [comment]
@eduard13 , feedback implemented.

**Comment by eduard13** [comment]
@magento run all tests

**Comment by sky-hub** [comment]
@magento run Integration Tests, Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by sky-hub** [comment]
@magento run Functional Tests B2B

**Review by eduard13 — APPROVED** [review]
Good job 👍

**Comment by eduard13** [comment]
The failed test doesn't seem to be related with the current PR.

**Comment by magento-engcom-team** [comment]
Hi @eduard13, thank you for the review.
[ENGCOM-8493](https://jira.corp.magento.com/browse/ENGCOM-8493) has been created to process this Pull Request

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Install Magento 2.4 with sample data;
2. Create cart rule with fixed amount discount  for whole cart (Ex: 50$);
3. Add a bundle product to cart (Example: http://local.magento2.com/sprite-yoga-companion-kit.html)
4. Navigate to checkout/cart page and check discount.

**Before:** :heavy_multiplication_x: Cart discount is 18.85$



**After:** :heavy_check_mark: Cart discount is 50$



Also was checked with **fixed amount discount** and everything works as expected.

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests CE

**Pull request merged**